### PR TITLE
TG Statue Fixes/Refactor

### PR DESCRIPTION
**Pull request by Fox-McCloud**
TG's Statue Fix/Refactor

Implement's TG's version of statues.

Basically, they have the same spells as our statues, but they're impervious to all damage (except gibbing). That said, they have a very very strict set of rules that governs if they can attack or not---if it's pitch black, then the statue can move/attack. If it's not, then the statue may only attack if everyone that has eyes on them (a mob with a client that is within vision range) is blinded. Again, if a mob is blinded, then that mob can be attacked, regardless of brightness.

Summary of changes

- Statues are now invincible to anything other than the most devastating of bombs
- Statues can no longer move or attack unless everyone around them is blind or in total darkness
- Statue light flicker range increased by 2
- Melee attack damage doubled
- Blindness spell no longer stuns silicons.

Fixes

- Fixes a bug where statues could typically disobey their lighting/blindness rules and just attack you anyway.
- Statues no long blind themselves with the blindness spell

Other
- Implements has_unlimited_silicon_privilege; largely unused for now, but can be utilized for other things later (only implemented it because ti was used in the statue PR).

:cl: Fox McCloud
tweak: Statues can no longer move/attack people unless they're in total darkness or all mobs viewing them are blinded
tweak: statues are invincible to anything other than full out gibbing.
tweak: Statues melee damaged doubled
tweak: Statues flicker lights spell range increased
tweak: Blind spell no longer impacts silicons
fix: Fixes statues being able to attack/move whenever they feel like it, regardless of client statues
fix: Fixes blind spell impacting the statue
/:cl:

### CODE REVIEW

File: code/modules/mob/living/simple_animal/hostile/statue.dm [hunk 1]
```diff
@@ -49,9 +52,9 @@
 /mob/living/simple_animal/hostile/statue/New(loc, var/mob/living/creator)
 	..()
 	// Give spells
-	AddSpell(new /obj/effect/proc_holder/spell/aoe_turf/flicker_lights(src))
```
Comment (tigercat2000) [inline code/modules/mob/living/simple_animal/hostile/statue.dm]: as mentioned these should be restored

File: code/modules/mob/living/simple_animal/hostile/statue.dm [hunk 2]
```diff
@@ -49,9 +52,9 @@
...
-	AddSpell(new /obj/effect/proc_holder/spell/aoe_turf/blindness(src))
-	AddSpell(new /obj/effect/proc_holder/spell/targeted/night_vision(src))
+	spell_list += new /obj/effect/proc_holder/spell/aoe_turf/flicker_lights(src)
```
Comment (Tastyfish) [inline code/modules/mob/living/simple_animal/hostile/statue.dm]: :eyes:

File: code/modules/mob/living/simple_animal/hostile/statue.dm [hunk 3]
```diff
@@ -60,137 +63,95 @@
...
-	//This loop will, at most, loop twice.
+	// This loop will, at most, loop twice.
 	for(var/atom/check in check_list)
 		for(var/mob/living/M in viewers(world.view + 1, check) - src)
-			if(M.client && CanAttack(M))
-				if(M.blinded || (M.sdisabilities & BLIND))
-					return null
-
-				//calculates if any mobs are actually facing the statue
-				var/xdif = M.x - src.x
```
Comment (tigercat2000) [inline code/modules/mob/living/simple_animal/hostile/statue.dm]: you never remade this in CanAttack()

File: code/modules/mob/living/simple_animal/hostile/statue.dm [hunk 4]
```diff
@@ -60,137 +63,95 @@
...
-	if(!.)
-		return
-	spawn() //for saftey
-		checkTargets()
 
-/mob/living/simple_animal/hostile/statue/proc/checkTargets()
-	if(target)
+/mob/living/simple_animal/hostile/statue/process_ai()
+	..()
+	if(!client && target) // If we have a target and we're AI controlled
```
Comment (Tastyfish) [inline code/modules/mob/living/simple_animal/hostile/statue.dm]: process_ai() isn't called if there's a client. :wink:

File: code/modules/mob/living/simple_animal/hostile/statue.dm (hunk 3 above)
Comment (tigercat2000) [inline code/modules/mob/living/simple_animal/hostile/statue.dm]: guess if they can't move in light it doesn't matter

**Comment by HereticChurch** [comment]
I really don't think they need a melee damage buff. They already had one of the highest single-hit damages of all mobs, generally breaking bones on the first hit. This will... wow, make them basically two shot crit people. That's kind of absurd.

Granted, with lights working on them they will attack less, so it might be balanced? I donno, doesn't sit well having something that can do 80 damage in a swipe.

**Comment by Spacemanspark** [comment]
I'm heavily agreeing with HereticChurch.
And they don't need to be invincible either. That's just... no.

File: code/modules/mob/living/simple_animal/hostile/statue.dm (hunk 4 above)
Comment (Fox-McCloud) [inline code/modules/mob/living/simple_animal/hostile/statue.dm]: ahh, right-o.

so just `if(target)`

File: code/modules/mob/living/simple_animal/hostile/statue.dm [hunk 5]
```diff
@@ -60,137 +63,95 @@
...
-
-	if(destination && destination.dynamic_lighting)
-		DLlight_amount = destination.get_lumcount()
-	else
-		DLlight_amount = 5
-
-	if(T && destination)
-		//Don't check it twice if our destination is the tile we are on or we can't even get to our destination
+	if(T && destination && T.lighting_overlay)
+		if(T.get_lumcount() * 10 < 1 && destination.get_lumcount() * 10 < 1) // No one can see us in the darkness, right?
```
Comment (Tastyfish) [inline code/modules/mob/living/simple_animal/hostile/statue.dm]: Why not `if(T.get_lumcount() < 10 && destination.get_lumcount() < 10)`

Comment (Fox-McCloud) [inline code/modules/mob/living/simple_animal/hostile/statue.dm]: was largely following the standard practice elsewhere, where `get_lumcount` is utilized.

File: code/modules/mob/living/simple_animal/hostile/statue.dm (hunk 4 above)
Comment (Tastyfish) [inline code/modules/mob/living/simple_animal/hostile/statue.dm]: yeah

**Comment by Earthdivine** [comment]
Whats the point of a 80 damage invincible mob? Even from an adminboose standpoint that is completely overkill. And then it can be summoned in gold cores? Thats a station wipe asking to happen, because it takes the lights going out in Xenobio and they can get into the maintence tunnels. This will also screw over shadowling rounds too if it gets out during that.

**Comment by Fox-McCloud** [comment]
@Earthdivine 

They were spawnable, in the past, and weren't generally problematic---and that was with the restrictions on them that didn't even work, half the time (though they did do half the damage).

If it's a huge concern, then I can leave them be and they can stay wizard only (as they currently are).

**Comment by Mijhead** [comment]
Do statues float now when they enter space, or are they still completely incapable of movement if they happen to slip 1 tile away from a wall?

**Comment by Earthdivine** [comment]
@Fox-McCloud 

In the past they also could be killed, and as you said the half damage as well. Having them be wizard only is perfectly fine, as that's the wizard using 2 spell slots to obtain these mobs and needs to use them both at the same time. I'm more worried about a Xenobio spawning a statue during a different round type like Shadowling and having the AI let it out.

**Comment by Fox-McCloud** [comment]
I'll leave them out of Xenobio for now.

**Comment by TheDZD** [comment]
I'd leave the damage as it was before, humans on TG don't need to worry about getting bones broken or organs destroyed due to one hit from these.

**Comment by Zilenan91** [comment]
They instead had to worry about DYING in 2 hits from these things. You also seem to really overestimate just how powerful statues are, it's a simple matter to just wall them in and them never being a problem ever again. If you're one person against one statue you're going to die, so don't go alone.

File: code/modules/mob/living/simple_animal/hostile/statue.dm [hunk 6]
```diff
@@ -60,137 +63,95 @@
...
 			if(get_dist(watching, src) > get_dist(target, src))
 				LoseTarget()
 				GiveTarget(watching)
 
 /mob/living/simple_animal/hostile/statue/AttackingTarget()
-	if(!can_be_seen())
-		icon_state = "angelattack"
+	if(can_be_seen(get_turf(loc)))
+		if(client)
+			src << "<span class='warning'>You cannot attack, there are eyes on you!</span>"
```
Comment (tigercat2000) [inline code/modules/mob/living/simple_animal/hostile/statue.dm]: to_chat

File: code/modules/mob/living/simple_animal/hostile/statue.dm [hunk 7]
```diff
@@ -60,137 +63,95 @@
 	if(creator)
 		src.creator = creator
 
-/mob/living/simple_animal/hostile/statue/Move(var/turf/NewLoc)
+/mob/living/simple_animal/hostile/statue/Move(turf/NewLoc)
 	if(can_be_seen(NewLoc))
-		icon_state = "angelseen"
 		if(client)
 			src << "<span class='warning'>You cannot move, there are eyes on you!</span>"
```
Comment (tigercat2000) [inline code/modules/mob/living/simple_animal/hostile/statue.dm]: to_chat (probably auto-resolved with a merge)

**Pull request closed without merging**